### DISCUSSION TITLE
How to listen to  io.quarkus.scheduler.SuccessfulExecution and  io.quarkus.scheduler.FailedExecution

### DISCUSSION THREAD

**Discussion by kirankolli-addepar** [quarkusio/quarkus#38960] [Q&A] [answered]
I have a class as below where I would like to listen to successful and failed scheduled execution . However this does not work. Am I doing anything wrong ? 

```
@Singleton
public class Singleton{
public void postExecution(@Observes SuccessfulExecution execution){
    log.info("Post successful execution : {}", Objects.toString(execution));    
  }

  public void postExecution(@Observes FailedExecution execution){
    log.info("Post failed execution : {}", Objects.toString(execution));    
  }
}

```

The methods are never fired. Is this a supported feature ? How do I listen to these events ? 
Is there a guide explaining how one can listen to these events ? 

Also please do note that I do have methods annotated with `@Scheduled(every = "1s")` annotation and the methods do get triggered.

[a comment by an automated account is not shown]

**Comment by mkouba** [comment]
Those observers look fine but the important thing is that the events are only fired for `@Scheduled` methods and jobs that were [registered programmatically](https://quarkus.io/guides/scheduler-reference#programmatic_scheduling) with the `JobDefinition` API. What jobs are you trying to observe?

**Reply by kirankolli-addepar** [reply]
yup, I do have a Scheduled annotation annotated method in another class : 

```
@Scheduled(every = "1s")
  public void process() {
 
  }
```

**Reply by kirankolli-addepar** [reply]
the scheduled method does fire , but the listener methods do not.

**Reply by mkouba** [reply]
That's weird. Do you use the correct `jakarta.*` annotations?

**Reply by kirankolli-addepar** [reply]
no I am using the `io.quarkus.scheduler.Scheduled` annotation

**Reply by kirankolli-addepar** [reply]
If I was on 2.16.5.Final version, which annotation should I be using ?

**Reply by kirankolli-addepar** [reply]
these are my imports on the listener : 
```
import io.quarkus.scheduler.FailedExecution;
import io.quarkus.scheduler.ScheduledExecution;
import io.quarkus.scheduler.SuccessfulExecution;
import jakarta.enterprise.event.Observes;
import jakarta.inject.Singleton;
```

**Reply by mkouba** [reply]
> no I am using the io.quarkus.scheduler.Scheduled annotation

That's correct. I was talking about `@Observes` and `@Singleton`.

> If I was on 2.16.5.Final version, which annotation should I be using ?

Quarkus 3 switched to `jakarta` packages. So you should be using `javax.*`.

**Reply by kirankolli-addepar** [reply]
yup, so looks like my imports are correct but the listeners do not work.

**Reply by mkouba** [reply]
We'll need a simple reproducer app in order to find out more. In any case, this functionality is tested so there might be something wrong in your config.

**Reply by manovotn** [reply]
> these are my imports on the listener :
> 
> ```
> import jakarta.enterprise.event.Observes;
> import jakarta.inject.Singleton;
> ```

If you are using Quarkus 2.16.x, then the above two annotations should instead be:
```
import javax.enterprise.event.Observes;
import javax.inject.Singleton;
```

Here's a link to the a test asserting exactly this behavior on Quarkus 2.16 - https://github.com/quarkusio/quarkus/blob/2.16/extensions/scheduler/deployment/src/test/java/io/quarkus/scheduler/test/FailedExecutionTest.java

**Reply by kirankolli-addepar** [reply]
@mkouba and @manovotn one more point that I left out as I thought it would not affect the situation is that my Scheduled annotation also has `skipExecutionIf=SomeClass.class` . Would this cause a bug ?

**Reply by mkouba** [reply]
If the `SomeClass` skips the execution then no events are fired.

**Reply by manovotn** [reply]
That would just skip the execution (leading to no events fired) but you earlier said that you can see the scheduled method being triggered so I don't suppose that's the case.

**Reply by kirankolli-addepar** [reply]
yup, as I said the method is getting called so i do expect the events to be fired.

**Reply by manovotn** [reply]
@kirankolli-addepar please provide a minimal reproducer and we'll be happy to take a closer look. Otherwise it's really hard to say what's happenning.

**Reply by kirankolli-addepar** [reply, marked as answer]
@manovotn , @mkouba , I found the culprit to be a very long `Thread.sleep()` in the scheduled method which was preventing the method from finishing execution and hence was not firing events.

I discovered this when I was trying to create a reproducer project for you.

Apologies and thanks.
